cryptify-back-end/src/main.rs

```diff
@@ -0,0 +1,379 @@
+        }
+        .to_string();
+        let content_range = match request.headers().get_one("Content-Range") {
+            Some(content_range) => content_range,
+            None => {
+                return rocket::request::Outcome::Failure((
+                    rocket::http::Status::BadRequest,
+                    "Missing content range".into(),
+                ))
+            }
+        }
+        .parse::<ContentRange>();
+        let content_range = match content_range {
+            Ok(v) => v,
+            Err(e) => {
+                return rocket::request::Outcome::Failure((rocket::http::Status::BadRequest, e))
+            }
+        };
+
+        rocket::request::Outcome::Success(UploadHeaders {
+            cryptify_token,
+            content_range,
+        })
+    }
+}
+
+#[derive(Responder)]
+struct UploadResponder {
+    body: (),
+    cryptify_token: CryptifyToken,
+}
+
+fn bytes_to_hex(bytes: &[u8]) -> String {
+    let mut s = String::with_capacity(2 * bytes.len());
+    for byte in bytes {
+        write!(s, "{:02x}", byte).unwrap();
+    }
+    s
+}
+
+fn compute_hash(cryptify_token: &[u8], data: &[u8]) -> String {
+    let mut hash = sha2::Sha256::new();
+    hash.update(cryptify_token);
+    hash.update(data);
+    format!("{:x}", hash.finalize())
+}
+
+#[put("/fileupload/<uuid>", data = "<data>")]
+async fn upload_chunk(
+    config: &State<CryptifyConfig>,
+    store: &State<Store>,
+    uuid: &str,
+    headers: UploadHeaders,
+    data: Data<'_>,
+) -> Result<Option<UploadResponder>, Error> {
+    let state = match store.get(uuid) {
+        Some(v) => v,
+        None => return Ok(None),
+    };
+    let mut state = state.lock().await;
+
+    if uuid::Uuid::parse_str(uuid).is_err() {
+        return Ok(None);
+    }
+
+    let start = headers.content_range.start.ok_or_else(|| Error::BadRequest(Some(
+        "Could not read Content-Range start".to_owned(),
+    )))?;
+    let end = headers.content_range.end.ok_or_else(|| Error::BadRequest(Some(
+        "Could not read Content-Range start".to_owned(),
+    )))?;
+    if end - start > 1024 * 1024 {
+        return Err(Error::BadRequest(Some(
+            "File chunk too large; the maximum is 1 MB".to_owned(),
+        )));
+    }
+
+    if headers.cryptify_token != state.cryptify_token {
+        return Err(Error::BadRequest(Some(
+            "Cryptify Token header does not match".to_owned(),
+        )));
+    }
+
+    let mut file = match OpenOptions::new()
+        .write(true)
+        .open(Path::new(config.data_dir()).join(uuid))
+        .await
+    {
+        Ok(v) => v,
+        Err(_) => return Ok(None),
+    };
+
+    if start >= end || state.uploaded != start {
+        return Err(Error::BadRequest(Some(
+            "Incorrect Content-Range header".to_owned(),
+        )));
+    }
+
+    file.seek(std::io::SeekFrom::Start(start))
+        .await
+        .map_err(|_| Error::InternalServerError(Some("Could not write file".to_owned())))?;
```

Comment:
Couldn't you just use an .or instead of .map_err (as closure is unused)?

Reply:
Yes, but it is optimised by Clippy by:
https://rust-lang.github.io/rust-clippy/master/index.html#or_fun_call
And then:
https://rust-lang.github.io/rust-clippy/master/index.html#bind_instead_of_map